console-subscriber/tests/support/subscriber.rs

```diff
@@ -0,0 +1,318 @@
+                }
+            }
+        }))
+        .await
+        .expect("client-console error: couldn't create client");
+    test_state.advance_to_step(TestStep::ClientConnected);
+
+    record_actual_tasks(channel, test_state.clone()).await
+}
+
+/// Records the actual tasks which are received by the client channel.
+///
+/// Updates will be received until the test state reaches the `TestFinished` step
+/// (indicating that the test itself has finished running), at which point we wait
+/// for a final update before returning all the actual tasks which were recorded.
+///
+/// # Test State
+///
+/// 1. Waits for: `TestFinished`
+async fn record_actual_tasks(
+    client_channel: Channel,
+    mut test_state: TestState,
+) -> Vec<ActualTask> {
+    let mut client = InstrumentClient::new(client_channel);
+
+    let mut stream = loop {
+        let request = tonic::Request::new(InstrumentRequest {});
+        match client.watch_updates(request).await {
+            Ok(stream) => break stream.into_inner(),
+            Err(err) => panic!("Client cannot connect to watch updates: {err}"),
+        }
+    };
+
+    let mut tasks = HashMap::new();
+
+    let mut last_update = false;
+    while let Some(update) = stream.next().await {
+        match update {
+            Ok(update) => {
+                if let Some(task_update) = &update.task_update {
+                    for new_task in &task_update.new_tasks {
+                        if let Some(id) = &new_task.id {
+                            let mut actual_task = ActualTask::new(id.id);
+                            for field in &new_task.fields {
+                                if let Some(console_api::field::Name::StrName(field_name)) =
+                                    &field.name
+                                {
+                                    if field_name == "task.name" {
+                                        actual_task.name = match &field.value {
+                                            Some(Value::DebugVal(value)) => Some(value.clone()),
+                                            Some(Value::StrVal(value)) => Some(value.clone()),
+                                            _ => None, // Anything that isn't string-like shouldn't be used as a name.
+                                        };
+                                    }
+                                }
+                            }
+                            tasks.insert(actual_task.id, actual_task);
+                        }
+                    }
+
+                    for (id, stats) in &task_update.stats_update {
+                        if let Some(task) = tasks.get_mut(id) {
+                            task.wakes = stats.wakes;
+                            task.self_wakes = stats.self_wakes;
+                        }
+                    }
+                }
+            }
+            Err(e) => {
+                panic!("update stream error: {}", e);
+            }
+        }
```

Comment:
style nit, take it or leave it: i'm not a huge fan of the amount of "rightward drift" in this code, would something like this work instead?
```suggestion
    while let Some(update) = stream.next().await {
        let update = match update {
            Ok(u) => u,
            Err(e) => panic!("update stream error: {e}"),
        };

        if let Some(task_update) = &update.task_update {
            for new_task in &task_update.new_tasks {
                let id = match new_task.id {
                    Some(id) => id,
                    None => continue,
                };
                let mut actual_task = ActualTask::new(id.id);
                let name = new_task.fields.iter().find_map(|field| {
                    match field.name? {
                        console_api::field::Name::StrName(name) if name == "task.name" => Some(field.value.clone()),
                        _ => None,
                    }
                });
                actual_task.name = match name {
                    Some(Value::DebugVal(value)) => Some(value.clone()),
                    Some(Value::StrVal(value)) => Some(value.clone()),
                    _ => None, // Anything that isn't string-like shouldn't be used as a name.
                };
                tasks.insert(actual_task.id, actual_task);
            }

            for (id, stats) in &task_update.stats_update {
                if let Some(task) = tasks.get_mut(id) {
                    task.wakes = stats.wakes;
                    task.self_wakes = stats.self_wakes;
                }
            }
        }
```

Reply:
This is much nicer, thank you! I didn't see this comment until after further changes, so I couldn't apply it, but I copied it pretty much as is.